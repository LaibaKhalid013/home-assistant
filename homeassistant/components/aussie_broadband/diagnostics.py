"""Provides diagnostics for Aussie Broadband."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from homeassistant.components.diagnostics import async_redact_data

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant

TO_REDACT = ["address", "ipAddresses", "description", "discounts", "coordinator"]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    return {
        "services": [
            {
                "service": async_redact_data(service, TO_REDACT),
                "usage": async_redact_data(service["coordinator"].data, ["historical"]),
            }
            for service in hass.data[DOMAIN][config_entry.entry_id]["services"]
        ]
    }
