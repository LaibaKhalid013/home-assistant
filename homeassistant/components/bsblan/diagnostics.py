"""Diagnostics support for BSBLan."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant

    from . import HomeAssistantBSBLANData


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    data: HomeAssistantBSBLANData = hass.data[DOMAIN][entry.entry_id]
    return {
        "info": data.info.to_dict(),
        "device": data.device.to_dict(),
        "state": data.coordinator.data.to_dict(),
    }
