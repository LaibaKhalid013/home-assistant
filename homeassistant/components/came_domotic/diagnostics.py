"""Diagnostics support for Hue."""

from __future__ import annotations

from typing import Any

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

from .came_domotic_server import CameDomoticServer
from .const import DOMAIN


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    bridge: CameDomoticServer = hass.data[DOMAIN][entry.entry_id]
    if bridge.api_version == 1:
        # diagnostics is only implemented for V2 bridges.
        return {}
    # Hue diagnostics are already redacted
    return await bridge.api.get_diagnostics()
