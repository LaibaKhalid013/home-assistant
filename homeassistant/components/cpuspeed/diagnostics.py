"""Diagnostics support for CPU Speed."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from cpuinfo import cpuinfo

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    info: dict[str, Any] = cpuinfo.get_cpu_info()
    return info
