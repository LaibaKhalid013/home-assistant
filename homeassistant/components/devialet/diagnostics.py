"""Diagnostics support for Devialet."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .const import DOMAIN

if TYPE_CHECKING:
    from devialet import DevialetApi

    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    client: DevialetApi = hass.data[DOMAIN][entry.entry_id]

    return await client.async_get_diagnostics()
