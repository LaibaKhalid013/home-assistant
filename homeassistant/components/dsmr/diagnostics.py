"""Diagnostics support for DSMR."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from homeassistant.util.json import json_loads

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

    from . import DsmrConfigEntry


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: DsmrConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""

    return {
        "entry": {
            "data": {
                **config_entry.data,
            },
            "unique_id": config_entry.unique_id,
        },
        "data": json_loads(config_entry.runtime_data.telegram.to_json())
        if config_entry.runtime_data.telegram
        else None,
    }
