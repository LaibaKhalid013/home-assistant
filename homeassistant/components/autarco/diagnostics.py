"""Support for the Autarco diagnostics."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
from homeassistant.core import HomeAssistant

from . import AutarcoConfigEntry, AutarcoDataUpdateCoordinator

TO_REDACT = {CONF_EMAIL, CONF_PASSWORD}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: AutarcoConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    autarco_data: list[AutarcoDataUpdateCoordinator] = config_entry.runtime_data

    return {
        "config_entry_data": async_redact_data(dict(config_entry.data), TO_REDACT),
        "sites_data": [
            {
                "id": coordinator.site.site_id,
                "name": coordinator.site.system_name,
                "health": coordinator.site.health,
                "solar": {
                    "power_production": coordinator.data.solar.power_production,
                    "energy_production_today": coordinator.data.solar.energy_production_today,
                    "energy_production_month": coordinator.data.solar.energy_production_month,
                    "energy_production_total": coordinator.data.solar.energy_production_total,
                },
            }
            for coordinator in autarco_data
        ],
    }
