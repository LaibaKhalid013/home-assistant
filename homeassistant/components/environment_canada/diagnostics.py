"""Diagnostics support for Environment Canada."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_LATITUDE, CONF_LONGITUDE

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant

TO_REDACT = {CONF_LATITUDE, CONF_LONGITUDE}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinators = hass.data[DOMAIN][config_entry.entry_id]
    weather_coord = coordinators["weather_coordinator"]

    return {
        "config_entry_data": async_redact_data(dict(config_entry.data), TO_REDACT),
        "weather_data": dict(weather_coord.ec_data.conditions),
    }
