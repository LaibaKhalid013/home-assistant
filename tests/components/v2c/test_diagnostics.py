"""Test V2C diagnostics."""

from unittest.mock import AsyncMock

from syrupy import SnapshotAssertion

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

from . import init_integration

from tests.components.diagnostics import get_diagnostics_for_config_entry
from tests.typing import ClientSessionGenerator

# Fields to exclude from snapshot as they change each run
TO_EXCLUDE = {
    "id",
    "device_id",
    "via_device_id",
    "last_updated",
    "last_changed",
    "last_reported",
}


def limit_diagnostic_attrs(prop, path) -> bool:
    """Mark attributes to exclude from diagnostic snapshot."""
    return prop in TO_EXCLUDE


async def test_entry_diagnostics(
    hass: HomeAssistant,
    mock_config_entry: ConfigEntry,
    mock_v2c_client: AsyncMock,
    hass_client: ClientSessionGenerator,
    snapshot: SnapshotAssertion,
) -> None:
    """Test config entry diagnostics."""

    await init_integration(hass, mock_config_entry)

    assert await get_diagnostics_for_config_entry(
        hass, hass_client, mock_config_entry
    ) == snapshot(exclude=limit_diagnostic_attrs)
