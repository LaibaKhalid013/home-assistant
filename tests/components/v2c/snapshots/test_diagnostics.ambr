# serializer version: 1
# name: test_entry_diagnostics
  dict({
    'config_entry': dict({
      'data': dict({
        'host': '**REDACTED**',
      }),
      'disabled_by': None,
      'domain': 'v2c',
      'entry_id': 'da58ee91f38c2406c2a36d0a1a7f8569',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'EVSE 1.1.1.1',
      'unique_id': 'ABC123',
      'version': 1,
    }),
    'data': "<MagicMock name='Trydan().data' id='4511132720'>",
    'host_status': 200,
    'raw_data': dict({
      'BatteryPower': 0.0,
      'ChargeEnergy': 1.8,
      'ChargePower': 1500.27,
      'ChargeState': 2,
      'ChargeTime': 4355,
      'ContractedPower': 4600,
      'Dynamic': 0,
      'DynamicPowerMode': 2,
      'FVPower': 0.0,
      'FirmwareVersion': '2.1.7',
      'HousePower': 0.0,
      'ID': 'ABC123',
      'Intensity': 6,
      'Locked': 0,
      'MaxIntensity': 16,
      'MinIntensity': 6,
      'PauseDynamic': 0,
      'Paused': 0,
      'ReadyState': 0,
      'SlaveError': 4,
      'Timer': 0,
    }),
  })
# ---
